### PR TITLE
Add `telemetry flush` command to send telemetry events

### PR DESCRIPTION
**Pull request by erikareads**
Up until now we've only been collecting telemetry events in memory and not actually sending them to the telemetry server. This adds the actual send logic, which happens in a subprocess so that the main process can operate without interruption.

### CODE REVIEW

File: .changeset/chilly-dryers-design.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+---
+'vercel': major
+---
+
+Vercel CLI now Collects Telemetry Data
+
+The Vercel CLI collects telemetry data about general usage.
+
+Telemetry allows us to accurately gauge Vercel CLI feature usage, pain points, and customization across all customers. This data enables us to better tailor the Vercel CLI to our customers’ needs, support its continued growth relevance, and optimal developer experience, as well as verify if improvements are enhancing the baseline performance of all applications.
```
Comment (TooTallNate) [inline .changeset/chilly-dryers-design.md]:
```suggestion
Telemetry allows Vercel to accurately gauge Vercel CLI feature usage, pain points, and customization across all customers. This data enables us to better tailor the Vercel CLI to our customers’ needs, support its continued growth relevance, and optimal developer experience, as well as verify if improvements are enhancing the baseline performance of all applications.
```

File: .changeset/chilly-dryers-design.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+---
+'vercel': major
+---
+
+Vercel CLI now Collects Telemetry Data
+
+The Vercel CLI collects telemetry data about general usage.
```
Comment (TooTallNate) [inline .changeset/chilly-dryers-design.md:7]:
```suggestion
The Vercel CLI collects anonymized telemetry data about general usage.
```

File: packages/cli/scripts/build.mjs [hunk 3]
```diff
@@ -1,7 +1,8 @@
-import { join } from 'node:path';
+import path, { join } from 'node:path';
```
Comment (TooTallNate) [inline packages/cli/scripts/build.mjs]:
```suggestion
import { join } from 'node:path';
```

File: packages/cli/scripts/build.mjs [hunk 4]
```diff
@@ -30,10 +31,25 @@ await compileDevTemplates();
...
-});
+await Promise.all([
+  esbuild({
+    bundle: true,
+    external: externals,
+  }),
+  // We call send-telemetry as a subprocess, so it needs to be bundled separately
+  esbuild({
+    entryPoints: [
+      path.join(
```
Comment (TooTallNate) [inline packages/cli/scripts/build.mjs]:
```suggestion
      join(
```

File: packages/cli/scripts/build.mjs [hunk 5]
```diff
@@ -30,10 +31,25 @@ await compileDevTemplates();
 const pkgPath = join(process.cwd(), 'package.json');
 const pkg = JSON.parse(readFileSync(pkgPath, 'utf8'));
 const externals = Object.keys(pkg.dependencies || {});
-await esbuild({
-  bundle: true,
-  external: externals,
-});
+await Promise.all([
```
Comment (TooTallNate) [inline packages/cli/scripts/build.mjs]: Is this change still necessary? I thought we decided on implementing as a CLI subcommand?

File: packages/cli/src/util/telemetry/send-telemetry.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
+/**
```
Comment (TooTallNate) [inline packages/cli/src/util/telemetry/send-telemetry.ts]: Is this file still necessary? I thought we decided on implementing as a CLI subcommand?

File: utils/build.mjs [hunk 7]
```diff
@@ -36,11 +36,12 @@ export async function esbuild(opts = {}, cwd = process.cwd()) {
     throw new Error(`Failed to load "${configPath}`);
   }
 
-  const entryPoints = opts.bundle
-    ? // When bundling, assume `src/index.ts` is the entrypoint file
-      [path.join(cwd, 'src/index.ts')]
-    : // When not bundling, compile all files referenced by the `tsconfig.json` file
-      tsconfig.fileNames;
+  const entryPoints =
+    (opts.entryPoints ?? opts.bundle)
```
Comment (TooTallNate) [inline utils/build.mjs]: The parens here seem incorrect. But also, is this change necessary (assuming the other point about the CLI build script change not being necessary)?

**Review by TooTallNate — CHANGES_REQUESTED** [review]

**Comment by TooTallNate** [comment]
Let's add a more proper PR title/description too please 🙏

File: utils/build.mjs (hunk 7 above)
Comment (jeffsee55) [reply utils/build.mjs]: Yes I had planned to remove this in another PR as a follow-up. Happy to just do that here.

File: .changeset/chilly-dryers-design.md [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+Telemetry allows Vercel to accurately gauge Vercel CLI feature usage, pain points, and customization across all customers. This data enables us to better tailor the Vercel CLI to our customers’ needs, support its continued growth relevance, and optimal developer experience, as well as verify if improvements are enhancing the baseline performance of all applications.
+
+Participation in this program is optional, and you may opt-out if you would prefer not to share any telemetry information with Vercel.
+
+Learn more about Vercel CLI Telemetry.
```
Comment (TooTallNate) [inline .changeset/chilly-dryers-design.md]: Should this be a link to somewhere?

File: packages/cli/src/commands/telemetry/command.ts [hunk 9]
```diff
@@ -16,6 +16,16 @@ export const enableSubcommand = {
   examples: [],
 } as const;
 
+export const flushSubcommand = {
+  name: 'flush',
+  aliases: [],
+  description: 'Interal command to flush telemetry events',
```
Comment (TooTallNate) [inline packages/cli/src/commands/telemetry/command.ts]:
```suggestion
  description: 'Internal command to flush telemetry events',
```

File: packages/cli/src/util/telemetry/commands/telemetry/index.ts [hunk 10]
```diff
@@ -25,4 +25,7 @@ export class TelemetryTelemetryClient
     // NOTE: this function is intentionally not implemented
     // because the user has explicitly opted out of telemetry
   }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  trackCliSubcommandFlush(_: string) {}
```
Comment (TooTallNate) [inline packages/cli/src/util/telemetry/commands/telemetry/index.ts]: A comment explaining why this is empty would be nice.

File: packages/cli/src/util/telemetry/index.ts [hunk 11]
```diff
@@ -227,7 +230,79 @@ export class TelemetryEventStore {
...
+
+        childProcess.stderr.on('data', data => output.debug(data.toString()));
+        childProcess.stdout.on('data', data => output.debug(data.toString()));
+        childProcess.on('error', d => {
+          output.debug(d);
+        });
+
+        setTimeout(() => {
+          childProcess.kill();
+        }, 2000);
```
Comment (TooTallNate) [inline packages/cli/src/util/telemetry/index.ts]: What is this timeout here for? If it's intentional, let's add an explanation comment

File: packages/cli/src/index.ts [hunk 12]
```diff
@@ -786,8 +798,10 @@ const main = async () => {
     return 1;
   }
 
-  // specifically don't await this, we want to fire and forget
-  telemetryEventStore.save();
+  // The telemetry flush event is called by `telemetryEventStore.save`, and it reinvokes the `main` function
+  if (subSubCommand !== 'flush') {
+    await telemetryEventStore.save();
```
Comment (TooTallNate) [inline packages/cli/src/index.ts:803]: This seems brittle. How about spawning the subprocess with the disable telemetry env var instead?

File: packages/cli/test/unit/util/telemetry-check-status.test.ts [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+        config: {
+          telemetry: {
+            enabled: true,
+          },
+        },
+      });
+    });
+
+    it('does not inform the user', async () => {
+      await expect(client.stderr).not.toOutput(
```
Comment (TooTallNate) [inline packages/cli/test/unit/util/telemetry-check-status.test.ts:57]: `not.toOutput` relies on a (2 second?) timeout, and is generally discouraged from being used. Might be faster to buffer the entire stream's output and do some assertions on there, or use `createLineIterator()` like a few other tests do.

File: packages/cli/src/index.ts (hunk 12 above)
Comment (jeffsee55) [reply packages/cli/src/index.ts:803]: Agreed that it feels a bit brittle. We do have a `enabled` check inside the `.save` function as things are right now so an env var would prevent us from sending. 

https://github.com/vercel/vercel/pull/12555/files#diff-2d5a21aecc56e19e130f744d9c268cf35ff1de827416e704d27483a01ffbd94fR232

**Review by TooTallNate — APPROVED** [review]
